wandb/sdk/internal/context.py

```diff
@@ -0,0 +1,79 @@
+"""Context Keeper."""
+
+import threading
+from typing import Dict, Optional
+
+from wandb.proto.wandb_internal_pb2 import Record, Result
+
+
+class Context:
+    _cancel_event: threading.Event
+    _debug_record: Optional[Record]
+
+    def __init__(self) -> None:
+        self._cancel_event = threading.Event()
+        self._debug_record = None
+
+    def cancel(self) -> None:
+        self._cancel_event.set()
+
+    @property
+    def cancel_event(self) -> threading.Event:
+        return self._cancel_event
+
+
+def context_id_from_record(record: Record) -> str:
+    context_id = record.control.mailbox_slot
+    return context_id
+
+
+def context_id_from_result(result: Result) -> str:
+    context_id = result.control.mailbox_slot
+    return context_id
+
+
+class ContextKeeper:
+    _active_items: Dict[str, Context]
+
+    def __init__(self) -> None:
+        self._active_items = {}
+
+    def add_from_record(self, record: Record) -> Optional[Context]:
+        context_id = context_id_from_record(record)
+        if not context_id:
+            return None
+        context_obj = self.add(context_id)
+
+        # TODO: add debug setting?
```

Comment:
address todo

Reply:
agreed, this should be a setting, otherwise this could grow uncontrollably (in terms of memory size) in the pathological case when the sender never gets a chance to send stuff and clean up afterwards.